near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1300 @@
+        self.lft.clear();
+        self.rgt.clear();
+        self.key.clear();
+        self.val.clear();
+        self.len = 0;
+        self.root = 0;
+        env::storage_remove(&self.tree_prefix);
+    }
+
+    pub fn height(&self) -> u64 {
+        self.ht.get(&self.root).unwrap_or_default()
+    }
+
+    pub fn contains_key(&self, key: &K) -> bool {
+        self.val.get(key).is_some()
+    }
+
+    pub fn get(&self, key: &K) -> Option<V> {
+        self.val.get(key)
+    }
+
+    pub fn insert(&mut self, key: K, val: V) -> Option<V> {
+        if self.contains_key(&key) {
+            // key is already present, changing only associated value
+            self.val.insert(&key, &val)
+        } else {
+            let at = self.root;
+            let id = self.len;
+            let root = self.insert_at(at, id, &key);
+            self.set_root(root);
+            self.len += 1;
+            self.val.insert(&key, &val)
+        }
+    }
+
+    pub fn remove(&mut self, key: K) -> Option<V> {
+        if self.contains_key(&key) {
+            let root = self.do_remove(&key);
+            self.set_root(root);
+            self.len -= 1;
+            self.val.remove(&key)
+        } else {
+            // no such key, nothing to do
+            None
+        }
+    }
+
+    pub fn min(&self) -> Option<K> {
+        self.min_at(self.root)
+    }
+
+    pub fn max(&self) -> Option<K> {
+        self.max_at(self.root)
+    }
+
+    pub fn floor(&self, key: &K) -> Option<K> {
+        self.floor_at(self.root, key)
+    }
+
+    pub fn ceil(&self, key: &K) -> Option<K> {
+        self.ceil_at(self.root, key)
+    }
+
+    pub fn iter<'a>(&'a self) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::asc(&self).into_iter()
+    }
+
+    pub fn iter_from<'a>(&'a self, key: K) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::asc_from(&self, key).into_iter()
+    }
+
+    pub fn iter_rev<'a>(&'a self) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::desc(&self).into_iter()
+    }
+
+    pub fn iter_rev_from<'a>(&'a self, key: K) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::desc_from(&self, key).into_iter()
+    }
+
+    pub fn to_vec(&self) -> Vec<(K, V)> {
+        self.iter().collect()
+    }
+
+    //
+    // Internal utilities
+    //
+
+    fn set_root(&mut self, root: u64) {
+        env::storage_write(&self.tree_prefix, &serialize(&root));
+        self.root = root;
+    }
+
+    fn min_at(&self, mut at: u64) -> Option<K> {
+        loop {
+            match self.lft.get(&at) {
+                Some(lft) => at = lft,
+                None => break
+            }
+        }
+        self.key.get(&at)
+    }
+
+    fn max_at(&self, mut at: u64) -> Option<K> {
+        loop {
+            match self.rgt.get(&at) {
+                Some(rgt) => at = rgt,
+                None => break
+            }
+        }
+        self.key.get(&at)
+    }
+
+    fn floor_at(&self, mut at: u64, key: &K) -> Option<K> {
```

Comment:
From the code it appears that `floor_at` returns the key that is strictly smaller or `None` if there's no key smaller, and `ceil_at` returns the key which is strictly larger than `K` and `None` if there's no key larger.

It is more common to have `lower_bound` and `upper_bound`, where `lower_bound` returns the smallest key that is greater or equal, and `upper_bound` returns the smallest key that is strictly greater.

The current `ceil_at` already behaves as `upper_bound`. Can we add `upper_bound` as an alias for `ceil_at` and `lower_bound` that matches the description above?

And then expose corresponding iterators (iterate from lower_bound, iterate from upper_bound)

Reply:
ack

Reply:
done